peps/pep-0740.rst

```diff
@@ -205,40 +177,224 @@ changes to it:
+  representing an individual attestation.
+* Each attestation object **MUST** be verifiable by the index. If the index fails
   to verify any attestation in ``attestations``, it **MUST** reject the upload.
+  The format of attestation objects is defined under :ref:`attestation-object`
+  and the process for verifying attestations is defined under
+  :ref:`attestation-verification`.
 
-In addition to the above, the index **SHALL** enforce a consistency
-policy for release attestations via the following:
+Index changes
+-------------
+
+Simple Index
+^^^^^^^^^^^^
 
-* If the first file under a new release is supplied with ``attestations``,
-  then all subsequently uploaded files under the same release **MUST** also
-  have ``attestations``. Conversely, if the first file under a new release
-  does not have any ``attestations``, then all subsequent uploads under the
-  same release **MUST NOT** have ``attestations``.
-* All files under the same release **MUST** have the same set of well-known
-  attestation format keys.
+* When an uploaded file has one or more attestations, the index **MAY** include a
+  ``data-provenance`` attribute on its file link.
 
-The index **MUST** reject any file upload that does not satisfy these
-consistency properties.
+* When the ``data-provenance`` attribute is present, its value **SHALL** be
+  a base64 encoded :ref:`provenance object <provenance-object>`.
 
-Index changes
--------------
+JSON-based Simple API
+^^^^^^^^^^^^^^^^^^^^^
+
+* When an uploaded file has one or more attestations, the index **MAY** include a
+  ``provenance`` object in the ``file`` dictionary for that file.
+* ``provenance``, when present, **MUST** be a :ref:`provenance object <provenance-object>`.
+
+These changes require a version change to the JSON API:
+
+* The ``api-version`` must specify version 1.2 or later.
+
+.. _attestation-object:
+
+Attestation objects
+-------------------
+
+An attestation object is a JSON object with several required keys; applications
+or signers may include additional keys so long as all explicitly
+listed keys are provided. The required layout of an attestation
+object is provided as pseudocode below.
+
+.. code-block:: python
+
+  @dataclass
+  class Attestation:
+      version: Literal[1]
+      """
+      The attestation format's version, which is always 1.
+      """
+
+      verification_material: VerificationMaterial
+      """
+      Cryptographic materials used to verify `message_signature`.
+      """
+
+      message_signature: str
+      """
+      The attestation's signature, as `base64(raw-sig)`, where `raw-sig`
+      is the raw bytes of the signing operation.
+      """
+
+  @dataclass
+  class VerificationMaterial:
+      certificate: str
+      """
+      The signing certificate, as `base64(DER(cert))`.
+      """
+
+      transparency_entries: list[TransparencyLogEntry]
+      """
+      Zero or more transparency log entries for this attestation's signature
+      and certificate.
+      """
```

Comment:
I would also specify one or more signed timestamps to verify the certificate's validity (where signed timestamps can take multiple forms, RFC3161, Roughtime, a log entry, etc)

Reply:
Makes sense. I think specifying the shape of each of these might get tricky, however, especially for the ones that don't have a standard JSON layout 🙂 

Instead of specifying each, what do you think about language (roughly) like this?

> Each attestation object MUST come with one or more independent, signed sources of time for the attestation. Examples of this signed timestamp include transparency log entries, RFC 3161 timestamps, and Roughtime signatures. Verifiers SHOULD verify these timestamp sources as part of establishing the attestation's timeliness, but MAY choose not to for support or policy reasons.

That gives us flexibility in the "v1" attestation here, which we can then ratchet down to a "MUST" in a subsequent version (along with a better-defined set of the sources/layouts of signed time).

WDYT?

Reply:
I like it, the flexibility sounds good.
I'm also happy with stating "one" and not "zero" in this draft. I know we talked about the option to say zero if you understand the threat model, I think that can be left for a future revision that is focused on BYO PKI.